File: jev_log_analyzer/events.py
"""Bounded event parsing and best-effort credential redaction."""
import hashlib
import json
import re

MAX_LINE = 65536
MAX_EVENT = 16000
MAX_INPUT_BYTES = 32 * 1024 * 1024
ANSI = re.compile(r"\x1b(?:\[[0-?]*[ -/]*[@-~]|\][^\x07\x1b]*(?:\x07|\x1b\\))")
TIMESTAMP = re.compile(r"^\d{4}-\d\d-\d\d[T ][\d:.]+(?:Z|[+-]\d\d:\d\d)?[ \t]")
CONTINUATION = re.compile(r"^(?:\s+\S|Caused by:|Suppressed:|Traceback |[\w.]+(?:Error|Exception):|\s*\.\.\. \d+ more)")
SECRET_FIELD = re.compile(r"(?i)(password|passwd|secret|token|authorization|cookie|api[-_]?key|access[-_]?key(?:[-_]?id)?|private[-_]?key|credential)")
SECRET_VALUE = re.compile(r'''(?ix)(["']?(?:password|passwd|secret|(?:access_|refresh_)?token|api[-_]?key|access[-_]?key(?:[-_]?id)?|private[-_]?key|authorization|cookie|credentials?)["']?\s*[:=]\s*)("(?:\\.|[^"\\\n])*"|'(?:\\.|[^'\\\n])*'|[^\s,;]+)''')


def redact(text):
    text = ANSI.sub("", text)
    def clean(value):
        if isinstance(value, dict):
            return {k: "[REDACTED]" if SECRET_FIELD.search(k) else clean(v) for k, v in value.items()}
        if isinstance(value, list):
            return [clean(v) for v in value]
        return scrub(value) if isinstance(value, str) else value

    def scrub(value):
        value = re.sub(r"(?i)\b(Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+", r"\1 [REDACTED]", value)
        value = re.sub(r"(\b[a-zA-Z][a-zA-Z0-9+.-]{0,31}://)[^\s/@:]+:[^\s/@]+@", r"\1[REDACTED]@", value)
        value = re.sub(r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b", "[REDACTED]", value)
        return SECRET_VALUE.sub(lambda m: m[1] + '"[REDACTED]"', value)

    try:
        return json.dumps(clean(json.loads(text)), ensure_ascii=False)
    except (ValueError, TypeError, RecursionError):
        return scrub(text)


class StreamRedactor:
    """Suppress a PEM private-key block even when its lines arrive separately."""
    begin = re.compile(r"-----BEGIN (?:[A-Z0-9]+ )?PRIVATE KEY-----")
    end = re.compile(r"-----END (?:[A-Z0-9]+ )?PRIVATE KEY-----")

    def __init__(self):
        self.private_key = False

    def redact(self, text):
        if self.private_key or self.begin.search(text):
            self.private_key = not bool(self.end.search(text))
            return "[REDACTED PRIVATE KEY]"
        return redact(text)


def parse_stream(stream, source, max_events, max_bytes=MAX_INPUT_BYTES):
    events, warnings = [], []
    pending = None
    line_no = 0
    remaining = max_bytes
    redactor = StreamRedactor()
    checked_limit = False

    def read_line():
        nonlocal remaining, checked_limit
        if remaining <= 0:
            if not checked_limit and stream.read(1):
                warnings.append("Decompressed input byte limit reached; remaining input was not analyzed")
            checked_limit = True
            return b""
        raw = stream.readline(min(MAX_LINE + 1, remaining))
        remaining -= len(raw)
        return raw

    while True:
        raw = read_line()
        if not raw:
            break
        line_no += 1
        oversize = (len(raw) > MAX_LINE or remaining == 0) and not raw.endswith(b"\n")
        if oversize:
            while raw_tail := read_line():
                if raw_tail.endswith(b"\n"):
                    break
        text = raw.decode("utf-8", errors="replace").rstrip("\r\n")
        if not text:
            continue
        match = TIMESTAMP.match(text)
        timestamp = match[0].strip() if match else None
        body = text[match.end():] if match else text
        if pending and CONTINUATION.match(body):
            pending["line_end"] = line_no
            pending["line_count"] += 1
            combined = pending["text"] + "\n" + redactor.redact(body)
            pending["truncated"] |= oversize or len(combined) > MAX_EVENT
            pending["text"] = combined[:MAX_EVENT]
        else:
            if pending:
                events.append(pending)
            if len(events) >= max_events:
                warnings.append("Event limit reached; remaining input was not analyzed")
                pending = None
                break
            safe = redactor.redact(body)
            pending = dict(source=source, timestamp=timestamp, text=safe[:MAX_EVENT],
                           line_start=line_no, line_end=line_no, line_count=1,
                           truncated=oversize or len(safe) > MAX_EVENT)
    if pending:
        events.append(pending)
    if any(e["truncated"] for e in events):
        warnings.append("Oversized lines/events were truncated; their classification requires review")
    for event in events:
        event["id"] = hashlib.sha256(json.dumps([source, event["line_start"], event["text"]], sort_keys=True).encode()).hexdigest()[:20]
    return events, warnings


BASELINE_RULES = [
    {"id": "error_or_failure_keyword", "description": "Failure or resource-exhaustion keywords", "pattern": r"(?i)\b(error|fatal|panic|exception|critical|oomkilled|out of memory|crashloopbackoff)\b"},
    {"id": "http_5xx", "description": "HTTP status in the 500–599 range", "pattern": r'(?i)(?:status["\s:=]+|HTTP/\d(?:\.\d)?["\s]+)5\d\d\b'},
    {"id": "warning_level", "description": "Warning-level keywords", "pattern": r'(?i)\b(warn|warning)\b'},
]


def baseline(event):
    signals = [rule["id"] for rule in BASELINE_RULES if re.search(rule["pattern"], event["text"])]
    return {"important": bool(signals), "signals": signals}

File: src/events.rs
//! Byte-bounded framing, multiline events, and best-effort redaction before storage.
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::{collections::BTreeMap, sync::LazyLock};

pub const MAX_LINE: usize = 65_536;
pub const MAX_EVENT: usize = 16_000;
pub const MAX_INPUT: u64 = 32 * 1024 * 1024;
pub type Source = BTreeMap<String, Value>;
fn rx(pattern: &str) -> Regex {
    Regex::new(pattern).expect("constant regex")
}
static ANSI: LazyLock<Regex> =
    LazyLock::new(|| rx(r"\x1b(?:\[[0-?]*[ -/]*[@-~]|\][^\x07\x1b]*(?:\x07|\x1b\\))"));
static STAMP: LazyLock<Regex> =
    LazyLock::new(|| rx(r"^\d{4}-\d\d-\d\d[T ][\d:.]+(?:Z|[+-]\d\d:\d\d)?[ \t]"));
static CONT: LazyLock<Regex> = LazyLock::new(|| {
    rx(
        r"^(?:\s+\S|Caused by:|Suppressed:|Traceback |[\w.]+(?:Error|Exception):|\s*\.\.\. \d+ more)",
    )
});
static FIELD: LazyLock<Regex> = LazyLock::new(|| {
    rx(
        r"(?i)(password|passwd|secret|token|authorization|cookie|api[-_]?key|access[-_]?key(?:[-_]?id)?|private[-_]?key|credential)",
    )
});
static SECRET: LazyLock<Regex> = LazyLock::new(|| {
    rx(
        r#"(?ix)(["']?(?:password|passwd|secret|(?:access_|refresh_)?token|api[-_]?key|access[-_]?key(?:[-_]?id)?|private[-_]?key|authorization|cookie|credentials?)["']?\s*[:=]\s*)("(?:\\.|[^"\\\n])*"|'(?:\\.|[^'\\\n])*'|[^\s,;]+)"#,
    )
});
static AUTH: LazyLock<Regex> = LazyLock::new(|| rx(r"(?i)\b(Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+"));
static URL: LazyLock<Regex> =
    LazyLock::new(|| rx(r"(\b[a-zA-Z][a-zA-Z0-9+.-]{0,31}://)[^\s/@:]+:[^\s/@]+@"));
static JWT: LazyLock<Regex> =
    LazyLock::new(|| rx(r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b"));
static PEM: LazyLock<Regex> =
    LazyLock::new(|| rx(r"-----(BEGIN|END) (?:[A-Z0-9]+ )*PRIVATE KEY-----"));
static RULES: LazyLock<Vec<(&str, Regex)>> = LazyLock::new(|| {
    vec![
        (
            "error_or_failure_keyword",
            rx(
                r"(?i)\b(error|fatal|panic|exception|critical|oomkilled|out of memory|crashloopbackoff)\b",
            ),
        ),
        (
            "http_5xx",
            rx(r#"(?i)(?:status["\s:=]+|HTTP/\d(?:\.\d)?["\s]+)5\d\d\b"#),
        ),
        ("warning_level", rx(r"(?i)\b(warn|warning)\b")),
    ]
});

pub fn hash(value: &impl Serialize) -> String {
    let bytes = serde_json::to_vec(value).expect("serializable identity");
    format!("{:x}", Sha256::digest(bytes))[..20].to_owned()
}
pub fn strip_ansi(text: &str) -> String {
    ANSI.replace_all(text, "").into_owned()
}
pub fn console(text: &str) -> String {
    strip_ansi(text)
        .chars()
        .map(|c| if c.is_control() { ' ' } else { c })
        .collect()
}
fn scrub(text: &str) -> String {
    let text = AUTH.replace_all(text, "$1 [REDACTED]");
    let text = URL.replace_all(&text, "$1[REDACTED]@");
    let text = JWT.replace_all(&text, "[REDACTED]");
    SECRET.replace_all(&text, "${1}\"[REDACTED]\"").into_owned()
}
pub fn redact(text: &str) -> String {
    fn clean(value: &mut Value) {
        match value {
            Value::Object(map) => {
                for (key, value) in map {
                    if FIELD.is_match(key) {
                        *value = Value::String("[REDACTED]".into());
                    } else {
                        clean(value);
                    }
                }
            }
            Value::Array(items) => {
                for v in items {
                    clean(v);
                }
            }
            Value::String(s) => *s = scrub(s),
            _ => (),
        }
    }
    let text = strip_ansi(text);
    if let Ok(mut value) = serde_json::from_str::<Value>(&text) {
        clean(&mut value);
        value.to_string()
    } else {
        scrub(&text)
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Event {
    pub id: String,
    pub source: Source,
    pub timestamp: Option<String>,
    pub text: String,
    pub line_start: u64,
    pub line_end: u64,
    pub line_count: u64,
    pub truncated: bool,
    pub group_id: String,
    #[serde(flatten)]
    pub judgment: crate::jev::Judgment,
    pub baseline: Baseline,
    pub analysis_reused: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub analysis_representative_id: Option<String>,
}
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct Baseline {
    pub important: bool,
    pub signals: Vec<String>,
}
pub fn baseline(text: &str) -> Baseline {
    let signals: Vec<_> = RULES
        .iter()
        .filter(|(_, r)| r.is_match(text))
        .map(|(id, _)| (*id).to_owned())
        .collect();
    Baseline {
        important: !signals.is_empty(),
        signals,
    }
}

#[derive(Debug)]
pub struct Line {
    pub bytes: Vec<u8>,
    pub truncated: bool,
    pub private: bool,
}
/// Retains only MAX_LINE bytes. PEM scanning includes discarded bytes and split markers.
#[derive(Default)]
pub struct Framer {
    buffer: Vec<u8>,
    truncated: bool,
    private: bool,
    in_key: bool,
    marker: String,
}
impl Framer {
    pub fn push(&mut self, bytes: &[u8]) -> Vec<Line> {
        let mut lines = Vec::new();
        for chunk in bytes.split_inclusive(|b| *b == b'\n') {
            self.private |= self.in_key;
            let scan = format!("{}{}", self.marker, String::from_utf8_lossy(chunk));
            for m in PEM.captures_iter(&scan) {
                self.private = true;
                self.in_key = &m[1] == "BEGIN";
            }
            self.marker = scan
                .chars()
                .rev()
                .take(128)
                .collect::<String>()
                .chars()
                .rev()
                .collect();
            let room = MAX_LINE - self.buffer.len();
            self.buffer
                .extend_from_slice(&chunk[..chunk.len().min(room)]);
            self.truncated |= chunk.len() > room;
            if chunk.ends_with(b"\n")
                && let Some(line) = self.finish()
            {
                lines.push(line);
            }
        }
        lines
    }
    pub fn finish(&mut self) -> Option<Line> {
        if self.buffer.is_empty() && !self.truncated {
            return None;
        }
        let line = Line {
            bytes: std::mem::take(&mut self.buffer),
            truncated: self.truncated,
            private: self.private,
        };
        self.truncated = false;
        self.private = self.in_key;
        self.marker.clear();
        Some(line)
    }
}
/// Parser state is retained across stream reconnects, including private-key suppression.
pub struct Parser {
    source: Source,
    pending: Option<Event>,
    line: u64,
}
impl Parser {
    pub fn new(source: Source) -> Self {
        Self {
            source,
            pending: None,
            line: 0,
        }
    }
    pub fn feed(&mut self, line: Line) -> Option<Event> {
        self.line += 1;
        let raw = String::from_utf8_lossy(&line.bytes);
        let text = strip_ansi(raw.trim_end_matches(['\r', '\n']));
        let stamp = STAMP.find(&text);
        let timestamp = stamp.map(|m| m.as_str().trim().to_owned());
        let body = stamp.map_or(text.as_str(), |m| &text[m.end()..]);
        if body.is_empty() && !line.private {
            return None;
        }
        let safe = if line.private {
            "[REDACTED PRIVATE KEY]".to_owned()
        } else {
            redact(body)
        };
        if CONT.is_match(body)
            && let Some(pending) = self.pending.as_mut()
        {
            pending.line_end = self.line;
            pending.line_count += 1;
            pending.text.push('\n');
            pending.text.push_str(&safe);
            pending.truncated |= line.truncated || pending.text.len() > MAX_EVENT;
            truncate(&mut pending.text, MAX_EVENT);
            return None;
        }
        let completed = self.flush();
        let truncated = line.truncated || safe.len() > MAX_EVENT;
        let mut safe = safe;
        truncate(&mut safe, MAX_EVENT);
        self.pending = Some(Event {
            id: String::new(),
            source: self.source.clone(),
            timestamp,
            text: safe,
            line_start: self.line,
            line_end: self.line,
            line_count: 1,
            truncated,
            group_id: String::new(),
            judgment: Default::default(),
            baseline: Default::default(),
            analysis_reused: false,
            analysis_representative_id: None,
        });
        completed
    }
    pub fn flush(&mut self) -> Option<Event> {
        let mut event = self.pending.take()?;
        event.id = hash(&(
            &event.source,
            event.line_start,
            &event.timestamp,
            &event.text,
        ));
        event.group_id = crate::grouping::group_id(&event);
        event.baseline = baseline(&event.text);
        Some(event)
    }
    pub fn has_pending(&self) -> bool {
        self.pending.is_some()
    }
}
fn truncate(s: &mut String, max: usize) {
    let mut end = max.min(s.len());
    while !s.is_char_boundary(end) {
        end -= 1;
    }
    s.truncate(end);
}

#[cfg(test)]
mod tests {
    use super::*;
    pub fn parse(raw: &[u8]) -> Vec<Event> {
        let mut framer = Framer::default();
        let mut parser = Parser::new(Source::new());
        let mut events = vec![];
        for chunk in raw.chunks(13) {
            for line in framer.push(chunk) {
                events.extend(parser.feed(line));
            }
        }
        if let Some(line) = framer.finish() {
            events.extend(parser.feed(line));
        }
        events.extend(parser.flush());
        events
    }
    #[test]
    fn multiline_ids_and_unicode_bounds() {
        let raw = "2026-09-20T12:00:00Z ERROR failed\n  stack frame\nINFO ready\n";
        let e = parse(raw.as_bytes());
        assert_eq!(e.len(), 2);
        assert_eq!(e[0].line_count, 2);
        assert_eq!(e[0].timestamp.as_deref(), Some("2026-09-20T12:00:00Z"));
        assert_eq!(e[0].id, parse(raw.as_bytes())[0].id);
        let e = parse("😀".repeat(MAX_LINE).as_bytes());
        assert!(e[0].truncated);
        assert!(e[0].text.len() <= MAX_EVENT);
    }
    #[test]
    fn credentials_and_streamed_pem() {
        let raw = concat!(
            "\x1b[31mERROR password=fixture-value\x1b[0m\n-----BEGIN ENCRYPTED PRIVATE ",
            "KEY-----\nprivate-fixture\n-----END ENCRYPTED PRIVATE ",
            "KEY-----\nINFO ready\n"
        );
        let e = parse(raw.as_bytes());
        let text = serde_json::to_string(&e).unwrap();
        assert!(!text.contains("fixture-value"));
        assert!(!text.contains("private-fixture"));
        assert!(!text.contains("\\u001b"));
        assert!(text.contains("INFO ready"));
        let safe = redact(
            r#"{"nested":{"access_key_id":"fake-value","secret":{"x":"hidden"}},"a":"Bearer fake-value"}"#,
        );
        assert!(!safe.contains("fake-value"));
        assert!(!safe.contains("hidden"));
        assert!(!redact(r#"message password="a\"b" done"#).contains("a\\"));
    }
    #[test]
    fn key_marker_in_discarded_oversize_tail() {
        let marker = concat!(
            "-----BEGIN PRIVATE ",
            "KEY-----\nprivate-fixture\n-----END PRIVATE ",
            "KEY-----\nready"
        );
        let raw = format!("{}{marker}", "x".repeat(MAX_LINE + 7));
        let e = parse(raw.as_bytes());
        assert!(
            !serde_json::to_string(&e)
                .unwrap()
                .contains("private-fixture")
        );
        assert!(e[0].truncated);
    }
}
